### PR TITLE
VM Fix: Increase the performance of neo by 20%.

### PR DESCRIPTION
**Pull request by Jim8y**
# Description

This pr focuses on optimize the vm performance. After sampling the overhead of syncing the whole blockchain, i noticed that the clear operation on the zero_reference hash set consumes 16.5% of the general time. So I replace it by generate a new hash set object, leave the old object to the C# GC to clear.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

**Test Configuration**:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by roman-khimov** [comment]
Have you checked the overall sync time change? Likely it's not 20% and now GC has to do more work.

You can make VM several times faster by replacing VM "GC" with reference counting (see nspcc-dev/neo-go#2500 also).

**Comment by cschuchardt88** [comment]
I don't understand how that can make it faster. You are just making a memory leak for `GC` to come by and clean up. On a consensus this will eat tons of memory. At least set `zero_referred` to `null` so `GC` knows to clean it memory.

**Comment by Jim8y** [comment]
I have no idea why cause study the exact cause will take a very long time, but i get the bench data by profiling the chain syncing process. Will make more measurement.

**Comment by Jim8y** [comment]
Good news is that it indeed increase the performance, bad news is that it decrease the efficiency of storage IO. I guess might be caused by GC.
 
 @cschuchardt88  any way to avoid the clear operation without incur extra overhead?

**Comment by Jim8y** [comment]
Old

Screenshot 2024-01-16 at 17 44 36

New:

Screenshot 2024-01-16 at 17 45 06

File: src/Neo.VM/ReferenceCounter.cs [hunk 1]
```diff
@@ -81,7 +81,8 @@ internal int CheckZeroReferred()
         {
             if (zero_referred.Count > 0)
             {
-                zero_referred.Clear();
+                // Overhead: 16.5%
+                zero_referred = new(ReferenceEqualityComparer.Instance);//.Clear();
```
Comment (cschuchardt88) [inline src/Neo.VM/ReferenceCounter.cs:85]:
```suggestion
                zero_referred = null; // tell GC to collect
                zero_referred = new(ReferenceEqualityComparer.Instance);//.Clear();
```

**Comment by shargon** [comment]
Reuse the collection should avoid the internal collection's increase, isn't it?

**Comment by superboyiii** [comment]
> @shargon it would be great if this does not break the system. @superboyiii can you please help verify?

Sure.

**Comment by superboyiii** [comment]
@Jim8y Tested. All compatible.

**Review by shargon — COMMENTED** [review]
Seems that `Clear` is not very optimal:

https://github.com/dotnet/runtime/blob/3ae99a78d8967e5f19765c52f272febef5def623/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSet.cs#L202-L206

**Comment by shargon** [comment]
> Old
> 
> Screenshot 2024-01-16 at 17 44 36
> 
> New:
> 
> Screenshot 2024-01-16 at 17 45 06

Jimmy I see the new slower, can you create a Benchmark for it?

**Comment by cschuchardt88** [comment]
I dont think this should be merged. We need more information. For example, how you are measuring this? How you are sure with 100% certainty this actually fixes the issue and doesn't put the issue elsewhere?.

**Pull request closed without merging**